### PR TITLE
Cudatoolkit

### PR DESCRIPTION
**Pull request by scopatz**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml"
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [ ] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

Initial recipe for cudatoolkit.  This follows the permissions given at https://github.com/conda-forge/conda-forge.github.io/issues/1163

This was originally forked from https://github.com/AnacondaRecipes/cudatoolkit-feedstock.  However, I have done a lot of work here to streamline the build process, including:

* Moved downloads to the meta.yaml, to take advantage of conda-build's caching
* Changed configuration logic to live in subclasses, as it probably always should have.
* Cleaned up imports, coding style, etc.

### CODE REVIEW

**Comment by scopatz** [comment]
Happy to add anyone else who wants as a maintainer. CC @kkraus14 @jakirkham

File: recipes/cudatoolkit/meta.yaml
Comment (kkraus14) [inline recipes/cudatoolkit/meta.yaml:83]: Adding folks from the NVIDIA side who have historically helped / worked on this recipe and are in the loop with internal conversations:

```suggestion
    - scopatz
    - mike-wendt
    - jakirkham
    - raydouglass
    - kkraus14
```

File: recipes/cudatoolkit/meta.yaml
Comment (scopatz) [reply recipes/cudatoolkit/meta.yaml:68]: Conda-forge uses SPDX identifiers, so this is the equivalent identifier.  Is this an issue?

File: recipes/cudatoolkit/conda_build_config.yaml
Comment (kkraus14) [inline recipes/cudatoolkit/conda_build_config.yaml:4]: I don't think we need this image if we're maintaining the same approach of downloading and extracting the binary package

File: recipes/cudatoolkit/meta.yaml
Comment (scopatz) [inline recipes/cudatoolkit/meta.yaml:87]: @mike-wendt @jakirkham @raydouglass @kkraus14 - are you willing to be maintainers?

File: recipes/cudatoolkit/meta.yaml
Comment (kkraus14) [reply recipes/cudatoolkit/meta.yaml:68]: If the text doesn't end up being `NVIDIA End User License Agreement` and link to https://docs.nvidia.com/cuda/eula/index.html then it will be an issue.

Comment (scopatz) [reply recipes/cudatoolkit/meta.yaml:68]: OK, we'll just have to ignore the linter warnings then

Comment (scopatz) [reply recipes/cudatoolkit/meta.yaml:68]: @kkraus14, can you please explain the `and link to https://docs.nvidia.com/cuda/eula/index.html` part? 

* conda-build does not allow the `license_file` entry to be a URL (it has to be a file)
* The artifacts at https://anaconda.org/nvidia/cudatoolkit/files have `license_file: NVIDIA_EULA` entries, as in the recipe here.
* Happy to include the PDF as this entry (available at https://docs.nvidia.com/cuda/pdf/EULA.pdf) instead, if needed

Happy to do whatever is needed.

**Review by mike-wendt — APPROVED** [review]
LGTM thanks for doing this @scopatz

**Comment by jakirkham** [comment]
Thanks for picking this up Anthony! 😄

**Comment by jakirkham** [comment]
cc @JasonAtNvidia @gmarkall @leofang (for vis)

**Comment by scopatz** [comment]
Also, any ideas on how to get NvToolsExt on the windows CI machine? Do we need to download drivers and install them too?

**Comment by gmarkall** [comment]
What are your thoughts on maintaining multiple CUDA toolkit versions in conda-forge, e.g. 11.0, 11.1, 11.2, etc. as time goes on? The motivation behind this question is that when I added packaging of `cudadevrt` to the AnacondaRecipes/cudatoolkit-feedstock (https://github.com/AnacondaRecipes/cudatoolkit-feedstock/pull/5), it only got added for the latest CUDA version at the time (10.2) and not earlier versions - however, there was a request to add it to all of the cudatoolkit versions, rather than just the latest (https://github.com/AnacondaRecipes/cudatoolkit-feedstock/pull/5#issuecomment-641309451 from @leofang).

For reference, the Numba cudatoolkit conda recipe (https://github.com/numba/conda-recipe-cudatoolkit, now defunct / unmaintained) from which the AnacondaRecipes recipe was forked kept recipes for each toolkit in separate folders, which made it easier to go back and rebuild an old version of the toolkit with changes (rather than the recipes for older versions being "trapped" only in the git history).

**Comment by scopatz** [comment]
I think we'll go back and add previous versions and fill out the matrix. But that work can/should happen on the feedstock. I am just trying to get this through staged recipes with any version right now

File: recipes/cudatoolkit/meta.yaml
Comment (kkraus14) [reply recipes/cudatoolkit/meta.yaml]: In the conda web UI there's an option to change the URL, i.e. at https://anaconda.org/nvidia/cudatoolkit/settings/options one can set the License URL if they have the proper permissions.

This was a sticking point in getting permission unfortunately.

File: recipes/cudatoolkit/meta.yaml
Comment (leofang) [reply recipes/cudatoolkit/meta.yaml:88]: +1

File: recipes/cudatoolkit/meta.yaml
Comment (scopatz) [reply recipes/cudatoolkit/meta.yaml]: Ahh, I think that will have to happen once the feedstock is made and the package is uploaded then.

Comment (kkraus14) [reply recipes/cudatoolkit/meta.yaml]: Before merging it would be good to confirm that someone has the permission / ability to do this accordingly.

Comment (scopatz) [reply recipes/cudatoolkit/meta.yaml]: Here is the current list of conda-forge owners on anaconda.org:

* beckermr
* bgruening
* cjwright4242
* dougal
* ericdill
* fwaters
* isuruf
* jakirkham
* jjhelmus
* marcelotrevisani
* mbargull
* msarahan
* mvn
* mwcraig
* ocefpaf
* patricksnape
* pelson
* scastellarin
* scopatz
* sodre
* SylvainCorlay
* uwe.korn
* wolfv

So at lease @jakirkham and I are on there

File: recipes/cudatoolkit/build.py
Comment (gmarkall) [inline recipes/cudatoolkit/build.py:269]: At present this recipe has the same issue as https://github.com/AnacondaRecipes/cudatoolkit-feedstock/issues/12 - see also the PR to fix it at https://github.com/AnacondaRecipes/cudatoolkit-feedstock/pull/14. The change that would fix the issue here is:

```suggestion
                            if path.endswith('x64'):
                                if not Path(os.path.join(store, filename)).is_file():
                                    shutil.copy(os.path.join(path, filename), store)
```

Comment (scopatz) [reply recipes/cudatoolkit/build.py]: Thanks @gmarkall!

File: recipes/cudatoolkit/meta.yaml
Comment (leofang) [reply recipes/cudatoolkit/meta.yaml:88]:
```suggestion
    - kkraus14
    - leofang
```

**Comment by scopatz** [comment]
So I think for windows this is going to need a `choco install` command to get the drivers. https://chocolatey.org/packages/nvidia-display-driver This is somethings that is going to need to happen in the feedstock, probably with an update to conda-smithy.

Given that, I think that this is ready to be merged in, since all actual packages need to be built on the feedstock.

**Comment by scopatz** [comment]
CC @conda-forge/staged-recipes - this is ready for review

**Comment by scopatz** [comment]
Thanks @chrisburr!

**Comment by isuruf** [comment]
Heads up. When this goes live, all the conda-forge cuda downstream feedstocks and installs will start breaking because of channel_priority. (And I guess RAPIDS pipelines too)

What we should do is move the builds to `test` and then build all previous versions that we support and then relabel all of them at the same time.

**Comment by scopatz** [comment]
@isuruf - why do you think that channel priority will break downstream feedstocks.  That is not my understanding of how channel priority works. Perhaps I am missing something. Can you explain, please?

**Comment by scopatz** [comment]
Also my only goal for the time being is to get linux-ppc64le cudatoolkit 10.1 building, which is not on defaults

**Comment by isuruf** [comment]
This will add `cudatoolkit=11.0.3` build to conda-forge. So, if `conda-forge` channel is in higher priority, conda solver will not consider any `cudatoolkit` packages (not even 10.x) from `defaults.

**Comment by scopatz** [comment]
Even if the pin states it needs `>=10.2,<11`?

**Comment by isuruf** [comment]
Yes

**Comment by scopatz** [comment]
I mean my goal is to go back and add the 10 series so hopefully this won't take very long

**Comment by scopatz** [comment]
PRs welcome there!

**Comment by jakirkham** [comment]
+1 Happy to help

Was thinking we can just make this change on Anaconda.org directly. Once things are ready we can move it back pretty easily. Thoughts?

**Comment by scopatz** [comment]
Oh yeah, that works for me!

**Comment by jakirkham** [comment]
Package just got published for Linux. Have moved it to the `test` label for now.

Windows build appears to have failed. Provided some info in issue ( https://github.com/conda-forge/cudatoolkit-feedstock/issues/1 ).

**Comment by scopatz** [comment]
Thanks @jakirkham - yeah, I agree that we should move a lot of the discussion to feedstock, so thanks for doing that

**Pull request merged**
